### ISSUE TITLE
strict-type-predicates conflicts with type guards

### ISSUE THREAD

**Issue by ryanstaniforth** [palantir/tslint#3972]
### Bug Report

- __TSLint version__: 5.10
- __TypeScript version__: 2.9.2
- __Running TSLint via__: CLI

#### TypeScript code being linted

```ts
export const isBaseInput = (input: {}): input is BaseInput => {
    if (typeof (input as BaseInput).type !== 'string') {
        return false;
    }

    return true;
};
```

with `tslint.json` configuration:

```json
{
    "defaultSeverity": "error",
    "linterOptions": {
        "exclude": "./src/data/**"
    },
    "extends": "tslint:all",
    "rules": {
        "quotemark": [true, "single"],
        "typedef": [true, "call-signature", "arrow-call-signature", "parameter"],
        "arrow-return-shorthand": true,
        "no-magic-numbers": false,
        "interface-name": [true, "never-prefix"],
        "completed-docs": false
    }
}

```

#### Actual behavior
I receive the error `Expression is always false.`, from the `strict-type-predicates` rule for this line: `if (typeof (input as BaseInput).type !== 'string') {`

#### Expected behavior
While the actual behavior is technically correct, it conflicts with the common pattern of checking for existence/type of object keys in type guards. Is there another way around this or does this need fixing? I would propose the solution of allowing this behaving inside of a type guard. i.e. inside a function with the return signature of `x is X`.

**Comment by giladgray** [comment]
try typing the argument as `any`. i've found that to be a best practice for type guards: assume nothing on the input and enforce type in the return.

```ts
function isBlah(arg: any): arg is Blah;
```

**Comment by giladgray** [comment]
pretty sure 🔼 should fix the issue, as the compiler is actually correct that the expression would be false the way you've written it, as `{}` doesn't have a `type` field declared, so it can't be a string type.

**Comment by astorije** [comment]
I am having the same concerns with the following pattern:

```ts
interface Route {
  name: string;
  url: string;
}

interface Section {
  name: string;
  sections: ReadonlyArray<NavItem>;
}

type NavItem = Route | Section;

function isRoute(item: NavItem): item is Route {
  return (item as Route).url !== undefined;
}
```

which causes this rule to report:

```
ERROR: 14:10  strict-type-predicates  Expression is always true.
```

I "fixed" this by switching to:

```diff
-  return (item as Route).url !== undefined;
+  return (item as object).url !== undefined;
```

but I would prefer to keep `Route` (and be closer to the [docs recommendation](https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards)).
I would like not to resort to `any`, is there any better recommendation here? Thanks!

**Comment by astorije** [comment]
My bad actually, `item as object` passes the linter, but not the compiler obviously:

```ts
error TS2339: Property 'url' does not exist on type 'object'.
```

Right now we have decided to use the following, which makes both linter and compiler happy, though really not ideal:

```ts
return (item as Partial<Route>).url !== undefined;
```
